src/request/sign-swap/SignSwapApi.js

```diff
@@ -218,6 +257,80 @@ class SignSwapApi extends BitcoinRequestParserMixin(TopLevelApi) {
     }
 
+    /**
+     *
+     * @param {unknown} direction
+     * @returns {'left-to-right' | 'right-to-left'}
+     */
+    parseDirection(direction) {
+        if (typeof direction !== 'string' || (direction !== 'left-to-right' && direction !== 'right-to-left')) {
+            throw new Error('Invalid direction');
+        }
+        return direction;
+    }
+
+    /**
+     *
+     * @param {KeyguardRequest.PolygonTransactionInfo} request
+     * @param {string[]} allowedMethods
+     * @returns {PolygonOpenDescription
+     *     | PolygonOpenWithApprovalDescription
+     *     | PolygonRedeemDescription
+     *     | PolygonRedeemWithSecretInDataDescription}
+     */
+    parseOpenGsnForwardRequest(request, allowedMethods) {
+        request.request = this.parseOpenGsnForwardRequestRoot(request.request);
+
+        if (request.request.to !== CONFIG.USDC_HTLC_CONTRACT_ADDRESS) {
+            throw new Errors.InvalidRequestError('request.to address is not allowed');
+        }
+
+        const usdcHtlcContract = new ethers.Contract(
+            CONFIG.USDC_HTLC_CONTRACT_ADDRESS,
+            PolygonContractABIs.USDC_HTLC_CONTRACT_ABI,
+        );
+
+        // eslint-disable-next-line operator-linebreak
+        const description =
+            /** @type {PolygonOpenDescription
+             *     | PolygonOpenWithApprovalDescription
+             *     | PolygonRedeemDescription
+             *     | PolygonRedeemWithSecretInDataDescription}
+             */ (usdcHtlcContract.interface.parseTransaction({
+                data: request.request.data,
+                value: request.request.value,
+            }));
+
+        if (!allowedMethods.includes(description.name)) {
+            throw new Errors.InvalidRequestError('Requested Polygon contract method is invalid');
+        }
+
+        if (description.name === 'open' || description.name === 'openWithApproval') {
+            if (description.args.token !== CONFIG.USDC_CONTRACT_ADDRESS) {
+                throw new Errors.InvalidRequestError('Invalid USDC token contract in request data');
+            }
+
+            if (description.args.refundAddress !== request.request.from) {
+                throw new Errors.InvalidRequestError('USDC HTLC refund address must be same as sender');
+            }
+        }
+
+        if (description.name === 'redeem' || description.name === 'redeemWithSecretInData') {
+            if (description.args.target !== request.request.from) {
+                throw new Errors.InvalidRequestError('USDC HTLC target address must be same as sender');
+            }
+        }
+
+        // Also check that approval object exists when method is openWithApproval
+        if (description.name === 'openWithApproval') {
+            if (!request.approval) {
+                throw new Errors.InvalidRequestError('`approval` object required for openWithApproval method');
+            }
+        }
```

Comment:
```suggestion
        // Check that approval object exists when method is 'openWithApproval', and unset for other methods.
        if ((description.name === 'openWithApproval') !== !!request.approval) {
            throw new Errors.InvalidRequestError('`approval` object required for openWithApproval method and ignored for other methods');
        }
```
`approval` is required only for `openWithApproval` and ignored otherwise.